### PR TITLE
Add new entry menthod to GlobalCaches for memory monitor

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

To compute the size of an component, the memory monitor needs to get the size of the local branch of a component. For (node)groups, this is accomplished by calling a simple action on the entire (node)group. However, this will not work with the (Mutable)GlobalCache because it cannot run simple actions as there is no `simple_action` entry method for the (Mutable)GlobalCache. This PR adds a new entry method that will compute the size of the local branch of the (Mutable)GlobalCache and send it to the memory monitor. This action can only be called when the MemoryMonitor singleton is in the component list in the metavariables.

### Dependencies

Includes and depends on #3949. Only the most recent commit is for this PR.

### CODE REVIEW

File: tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp]: Sorry, since this uses a lock wouldn't it always be okay?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
LGTM, just confused on one comment

File: tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp
Comment (knelli2) [reply tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp]: I guess I'm not sure how file locks work. If the lock ensures that nothing is writing to/reading from the file, and that any writing that is occurring when the lock is called is finished first before the code continues after the lock, then I probably don't need this comment.

Comment (nilsdeppe) [reply tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp]: Yes, that's what the lock does. Of course someone could always write without using the lock, but that's incorrect usage of the H5 infrastructure :)

Comment (knelli2) [reply tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp]: Ok I just removed the comment.

**Comment by nilsdeppe** [comment]
LGTM, I'll approve once the PR this depends on is merged, just so I don't have a moment and accidentally merge this with the other PR not yet merged...

**Comment by knelli2** [comment]
@nilsdeppe Rebased on develop now that #3949 is merged.

**Review by nilsdeppe — DISMISSED** [review]

**Comment by knelli2** [comment]
Marking as "in progress" because I am unable to monitor the memory of the MutableGlobalCache in some BBH runs I'm doing because I get the following error:
```
------------- Processor 12 Exiting: Called CmiAbort ------------
Reason: Unrecognized PUP::able::PUP_ID. is there an unregistered module?
```
I'm able to monitor all other parallel components, even the GlobalCache. Just not the MutableGlobalCache. Possibly related to the fact that the `compute_size_for_memory_monitor` entry method for the mutable cache takes a proxy to the global cache.

**Comment by nilsdeppe** [comment]
This _usually_ comes from trying to serialize (or maybe size) an abstract base class that wasn't registered. Is there something in the MutableGlobalCache that was maybe not registered?

**Comment by knelli2** [comment]
Removing "in progess" label because I haven't found an easy fix for this. If it's ok with the reviewers (@nilsdeppe) I'd like to move on without fixing this issue and then we can revisit it later. I've opened #3993 about this so we can keep track of things. The new entry method for the MutableGlobalCache still works fine without control systems.

**Comment by nilsdeppe** [comment]
That's fine by me. I think the issue isn't in the code being added here, just that the code being added exposes the issue of something not being registered.

**Comment by nilsdeppe** [comment]
@wthrowe did you want to take a look at this too? Maybe you can comment if you'd like to review by end of day Wed. June 1 so I know to wait for your reviews?

File: src/Parallel/GlobalCache.hpp
Comment (markscheel) [inline src/Parallel/GlobalCache.hpp]: "This entry method of the MutableGlobalCache" -> "MutableGlobalCache::compute_size_for_memory_monitor"

File: src/Parallel/GlobalCache.hpp
Comment (markscheel) [inline src/Parallel/GlobalCache.hpp]: "This entry method of the GlobalCache" -> "GlobalCache::compute_size_for_memory_monitor"

File: tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp
Comment (markscheel) [inline tests/Unit/Parallel/Test_AlgorithmGlobalCache.cpp]: Isn't there only one branch since we are running on one core?

**Review by markscheel — COMMENTED** [review]
Looks good. Only a few clarifications in comments and messages.

Presumably, this will not report the size of any YlmSpherepacks in the global cache (e.g. inside of the Shape map) because YlmSpherepack is not puppable, correct?

**Comment by knelli2** [comment]
@markscheel Correct, at the moment this won't cover YlmSpherepack because it's not pupable. In the future I will probably add support for including things we don't `pup` in the memory monitor calculation.

**Review by markscheel — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
GCC 7 failure is `Unit.Evolution.Systems.Cce.Actions.ScriObserveInterpolated (Timeout)` ignoring.

**Pull request merged**
